### PR TITLE
Use all form submit fields when spidering

### LINKED ISSUE #2748: ZAP Spidering HTML Forms with multiple submit buttons
ZAP when encountering an HTML Form with multiple submit buttons defined

e.g.

```
<form action="test.jsp">
<input type="text" value="a" size="10" maxlength="10" name="name">
<input type="submit" value="Reload" name="submit1">
<input type="submit" value="Basic" name="submit2">
```

it spiders for the following URL
/test.jsp?name=a&submit1=Reload&submit2=Basic

instead of submitting two URLs like
/test.jsp?name=a&submit1=Reload
/test.jsp?name=a&submit2=Basic

Reference: https://github.com/zaproxy/zaproxy/blob/develop/src/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java#L109

Additionally, the "value" attribute of the input type="submit" element can be used as a differentiator in the edge case where the two submit buttons have the same ID/name

e.g.
```
<form action="test.jsp">
<input type="text" value="a" size="10" maxlength="10" name="name">
<input type="submit" value="Reload" name="submit">
<input type="submit" value="Basic" name="submit">
```

the spider can generate the following requests
/test.jsp?name=a&submit=Reload
/test.jsp?name=a&submit=Basic

### PR DESCRIPTION
**Pull request by thc202**
Change class SpiderHtmlFormParser to use all submit fields when
spidering, using each submit field in turn with remaining form fields.
Add some tests to assert the expected behaviour of SpiderHtmlFormParser.
Other minor changes were done to "normalise" the behaviour/expectations
of class SpiderHtmlFormParser:
 - Throw an exception when creating an instance with null SpiderParam,
 it's required to check if HTML POST forms should be parsed;
 - Change the instance variable "param" to private and final.
 - Expect always a message when parsing (letting a NullPointerException
 be thrown if not).

Fix #2748 - ZAP Spidering HTML Forms with multiple submit buttons

### CODE REVIEW

File: src/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java [hunk 1]
```diff
@@ -121,37 +121,39 @@ public boolean parseResource(HttpMessage message, Source source, int depth) {
...
 				}
 
-				log.debug("Canonical URL constructed using '" + action + "': " + fullURL);
-				notifyListenersPostResourceFound(message, depth + 1, fullURL, query);
+				for (HtmlParameter submitField : formData.getSubmitFields()) {
+					notifyPostResourceFound(
+							message,
+							depth,
+							fullURL,
+							appendEncodedUrlQueryParameter(baseRequestBody, submitField));
```
Comment (kingthorin) [inline src/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java]: Does this really need to be split across 5 lines?

File: test/org/zaproxy/zap/spider/parser/SpiderHtmlFormParserUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        SpiderHtmlFormParser htmlParser = createSpiderHtmlFormParser();
+        boolean parsed = true;
+        // When
+        boolean canParse = htmlParser.canParseResource(new HttpMessage(), ROOT_PATH, parsed);
+        // Then
+        assertThat(canParse, is(equalTo(false)));
+    }
+
+    @Test
+    public void shouldNotParseNonHtmlResponse() {
```
Comment (kingthorin) [inline test/org/zaproxy/zap/spider/parser/SpiderHtmlFormParserUnitTest.java]: This name seems a bit misleading `HttpMessage()` creates an html response, though without a filename param it might be empty. So it "is" an html response (per the set content type), there's just no content....

File: test/org/zaproxy/zap/spider/parser/SpiderHtmlFormParserUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        Source source = null;
+        SpiderHtmlFormParser htmlParser = createSpiderHtmlFormParser();
+        HttpMessage messageHtmlResponse = createMessageWith("NoForms.html");
+        // When
+        htmlParser.parseResource(messageHtmlResponse, source, BASE_DEPTH);
+        // Then = No exception
+    }
+
+    @Test
+    public void shouldNeverConsiderCompletelyParsed() {
```
Comment (kingthorin) [inline test/org/zaproxy/zap/spider/parser/SpiderHtmlFormParserUnitTest.java]: `shouldNeverConsiderCompletelyParsedWhenSourceNull`? Or why should this never be considered fully parsed?

**Comment by kingthorin** [comment]
Additional tests needed?
- POST test cases that mirror the included GET cases? 
- Single page with mixed form methods? (One GET one POST?)
- Form with no method specified? [_Should default to GET I guess?_]

File: src/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java (hunk 1 above)
Comment (thc202) [inline src/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java]: Depends, that's done to improve readability, but can change that.
(I'm using the formatter suggested in the dev mailing list a long time ago.)

File: test/org/zaproxy/zap/spider/parser/SpiderHtmlFormParserUnitTest.java (hunk 2 above)
Comment (thc202) [inline test/org/zaproxy/zap/spider/parser/SpiderHtmlFormParserUnitTest.java]: It's creating a new `HttpMessage` (but not using `createMessageWith`), which by default has no response (so, it's not HTML).

File: test/org/zaproxy/zap/spider/parser/SpiderHtmlFormParserUnitTest.java (hunk 3 above)
Comment (thc202) [inline test/org/zaproxy/zap/spider/parser/SpiderHtmlFormParserUnitTest.java]: Whether or not it's considered parsed does not depend on the inputs, that's why it just says that. Why it has that behaviour? I guess because it's a fall back parser.

**Comment by thc202** [comment]
Sure, will add those tests as well (I did test POST, though).

File: src/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java (hunk 1 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java]: Ok. Seems like overkill to me but it's a really minor detail.

File: test/org/zaproxy/zap/spider/parser/SpiderHtmlFormParserUnitTest.java (hunk 3 above)
Comment (kingthorin) [inline test/org/zaproxy/zap/spider/parser/SpiderHtmlFormParserUnitTest.java]: Ok

File: src/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java (hunk 1 above)
Comment (thc202) [inline src/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java]: I don't mind using other formatter rules that don't do that (using that one for consistency).

Comment (kingthorin) [inline src/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java]: Leave it for now.

Comment (thc202) [inline src/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java]: OK.

**Comment by psiinon** [comment]
@thc202 are you planning on adding those tests?

**Comment by thc202** [comment]
Yes.

**Comment by thc202** [comment]
Pull request updated with more tests. There's still some untested behaviour which will require some changes to normalise it (unrelated to this pull request, mostly about handling of HTML base element).

**Comment by thc202** [comment]
And it's rebased.

**Comment by kingthorin** [comment]
Looks good to me
